openassessment/tests/test_data.py

```diff
@@ -1319,6 +1319,58 @@ def test_create_zip_with_attachments(self):
 
         self.assertTrue(zip_file.read(os.path.join(COURSE_ID, 'downloads.csv')))
 
+    def test_create_zip_with_failed_attachments(self):
+        file = BytesIO()
+
+        with patch(
+            'openassessment.data.OraDownloadData._download_file_by_key'
+        ) as download_mock:
+            download_mock.side_effect = FileMissingException
+            OraDownloadData.create_zip_with_attachments(file, COURSE_ID, self.submission_files_data)
+
+            download_mock.assert_has_calls([
+                call(self.file_key_5),
+                call(self.file_key_4),
+                call(self.file_key_1),
+                call(self.file_key_2),
+                call(self.file_key_3),
+            ])
+
+        zip_file = zipfile.ZipFile(file)
+
+        # archive should contain only three parts text file and one csv because all of the attachments are invalid
+        self.assertEqual(len(zip_file.infolist()), 4)
+
+        # check for pre_file_name_user's text file
+        self.assertEqual(
+            zip_file.read(self.submission_files_data[1]['file_path']),
+            self.answer_text.encode('utf-8')
+        )
+        self.assertEqual(
+            zip_file.read(self.submission_files_data[3]['file_path']),
+            self.answer_text.encode('utf-8')
+        )
+
+        # main user's text response
+        self.assertEqual(
+            zip_file.read(self.submission_files_data[7]['file_path']),
+            self.answer_text.encode('utf-8')
+        )
+
+        # expect the submission file to be invalid
+        self.assertEqual(
```

Comment:
hmm, I didn't catch this till now, but I don't think we should be modifying the passed-in data dictionary. Do we have any tests that test the contents of the generated CSV file?

Reply:
this may be a harder ask than I originally intended, since there's not really any prior art on testing this CSV (the stuff at the top of the file is unrelated). If I pointed you at some libraries we use and some code we've written elsewhere, would you be down to try adding some tests for the content of the CSV? Other option: you could make another test where you call the CSV function with just like, one user and one or two uploaded files, and then just assert that the text of the CSV is equal to another static string you define in the test

Reply:
I will make a copy of data dictionary instead. Also, You can pointed the libraries and the code elsewhere. I do not have preference on this.

Reply:
changed to use spread instead of copy object from the original